nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+reference Id. This may be something that is built into the protocol or implemented ad-hoc. To support this at the
+protocol level, we would need to add the following APIs:
+
+- `Resolve(operation, input) -> reference_id`
+
+-- OR --
+
+- `Cancel(operation, input)` - cancel operation by handler derived Id
+- `GetResult(operation, input)` - get operation result by handler derived Id
+- `Describe(operation, input)` - get operation status by handler derived Id
+
+NOTE: This looks a lot like the APIs for the "client generated Ids" approach, except the ID is structured instead of a
+single string.
+
+#### Cancel before started
+
+A caller may cancel an operation before it obtains a reference Id, e.g. if it is uncertain that the operation has been
+already started due to crashes and timeouts. Because the caller doesn't have a referencable Id or knowledge of which
+input fields are used to derive the Id, it would need to pass in the entire Start call input.
+
+This is especially important when we want to automate cancellations of workflow initiated operations, Temporal server
+cannot derive the Id from opaque input.
+
+### Comparing the approaches
+
+#### Handler Id qualification
+
+With handler generated Ids, to support stateless clients, the handler would need to be able to map an operation and
+input to a reference Id. It would also need to qualify a reference Id or securely issue the reference Ids when callers
+cannot be trusted. With caller generated Ids, the handler would only need to qualify an Id based on the caller context
+(e.g by concatinating it with a tenant Id).
+
+#### Ids as inputs
+
+With the caller generated Id approach, if the Id is used for handler side addressability, the handler would need to
+parse it and merge with the request's input.
+
+#### Caller friendliness
+
+Shifting the responsibility for creating a handler addressable Id to the caller makes an API harder to use.
+
+If we can map arbitrary client Ids to handler addressable resources, the client generated Id approach is easiest to use
+from stateless clients. The SDK can generate deterministic Ids for calls started from workflows.
+
+Handler generated Ids is the least friendly approach for stateless clients without additional APIs for mapping input to
+a handler addressable resources.
+
+#### Data sensitivity
+
+If a client is expected to provide a business meaningful Id that encodes sensitive information to address an operation,
+the Id would need to be encrypted making it harder to use.
+
+Both handler generated Ids and caller generated Ids with handler mapping options avoid the data sensitivity issue.
+
+## Proposal
+
+- Require client generated Ids for any call that starts an asynchronous operation.
+- API will be based on the rough draft in [Caller generated identifiers](#caller-generated-identifiers).
+- Temporal exposes a strongly consistent mapping of caller to handler identifiers that handlers may choose to leverage.
```

Comment:
I was a bit torn here. Usually I would say allowing the clients to identify their operation they want to interact with statelessly is a good practice. It's the nice thing to do to let clients be stateless. This is what Temporal does with workflows and such. But looking around at https://google.aip.dev/151 and general I can understand that this is not resource identification, this is operation identification. If a handler wants to support a "create resource with ID" and then a "affect resource with ID" then that makes sense, but does that need to apply to operation? In traditional programming languages, one might expect a promise or future to be returned from an async op. The caller doesn't create the promise/future, rather the handler does and the caller is expected to use that promise/future.

I would like a hybrid approach of optional client identifier but handler generated if not present (i.e. how an auto generated sequence PK might be in a DB upon insert, but you can provide it) allowing the handler to remain stateless. But I can understand it's difficult to require clients to provide a _globally unique_ identifier while also letting them use a meaningful one.

So I could be convinced of either direction, but I slightly lean to the handler generated identifiers.

But maybe there's a hybrid we can take. Maybe there is some way we can uniquely identify clients same way as we do namespaces with workflows today? Then we can have client generated identifiers and tell handlers that we guarantee "client name + operation identifier are globally unique". Then both the client and handler can be stateless and the true identifier is a tuple of client and identifier.

> Temporal exposes a strongly consistent mapping of caller to handler identifiers that handlers may choose to leverage.

If we're really gonna support this, it needs to be more general purpose operation metadata the handler can set on start response and is delivered to the handler on each other invocation.

Reply:
> hybrid approach of optional client identifier but handler generated if not present ... allowing the handler to remain stateless. 

Does that actually allow the handler to remain stateless in the not-provided case? Regardless, I like that flexibility too.

Generally though, I agree with Chad.

>it's difficult to require clients to provide a globally unique identifier while also letting them use a meaningful one.

Is definitely true. Either you require some format like GUID, and rob clients of the opportunity to make the ID meaningful, or you don't, and have to spend cycles checking it's unique which could've been used to just return them an ID.

The analogy to futures makes sense to me as well

Reply:
> Does that actually allow the handler to remain stateless in the not-provided case? Regardless, I like that flexibility too.

Not if they need to translate to some internal ID. But if you take that namespace-ish identifier + client-provided operation identifier concatenated to be your workflow ID in the handler, then you can just reuse that for each operation on that workflow. But otherwise if there is a stateful translation needed, then yes some kind of meta store is needed.

Reply:
I think it is up to the implementer of the operation to map the operation arguments to the id. For example, it can return the same id for two separate calls if their arguments point to the same implementation workflow.

Reply:
I've been thinking about this some more and a few discussions resonated with me:

1. If a caller identifier is constrained to a single string, and is used to address a resource on the handler side, the caller may end up concatenating multiple identifiers to generate a single string, which the handler then has to parse. This is suboptimal experience IMHO.
2. Max's comment that with caller generated IDs, separate calls cannot map to the same workflow implementation.
3. Chad's comment (which has been mentioned in previous discussions), and the reference to Google's longrunning spec, where the caller doesn't "create" a promise, it can only get a reference to one from calling a method.
4. As I mentioned in this proposal, with caller generated IDs, extra care should be taken to encrypt any sensitive information that make up the ID.

The proposal to use client generated IDs came to me from writing this part of the spec:

> - `Cancel(operation, input)` - cancel operation by handler derived Id
> - `GetResult(operation, input)` - get operation result by handler derived Id
> - `Describe(operation, input)` - get operation status by handler derived Id
>
> NOTE: This looks a lot like the APIs for the "client generated IDs" approach, except the ID is structured instead of a single string.

Writing this made me think that we might as well go with caller generated IDs for stateless client support as it would be easier to implement the handler. All you would need is to map a caller context+ID to an internal resource ID, no need to sign or qualify reference IDs.
But this approach is less flexible. For full flexibility, I'm starting to think we'll want to expose methods for addressing an operation by both a handler provided reference ID and by an operation and structured input.

I'm now leaning more toward the API I mentioned in the handler generated IDs section:

```
Start(operation, input, ...)
CancelByID(reference_id)
GetResultByID(reference_id)
DescribeByID(reference_id)
// Option 1:
Cancel(operation, partial_input)
GetResult(operation, partial_input)
Describe(operation, partial_input)
// Option 2:
GetReferenceID(operation, partial_input)
```

A few things to note here:

1. To get a reference to an operation, the caller only needs to send the input fields that are needed to resolve the reference (very much like the caller generated IDs proposal, except this API doesn't require encoding input as string). In practice this input will often times be a single, caller unique, identifier.
2. The handler doesn't **have** to support resolving a reference from operation + input for all operation types, some APIs may not need to be idempotent (e.g. they may be read only). The handler will necessarily have the ability to map operation + input to an internal identifier (workflow ID) for any idempotent operation.
3. `Cancel` is equivalent to `GetReferenceID` + `CancelByID` but it requires one less round trip, I prefer that.
4. We can choose not to include either option 1 or option 2 in the API, and tell handler implementors to expose this functionality per use case with custom operations. For example a handler implementing a `charge` operation, would also implement `cancelCharge`, `describeCharge`, etc... but I would argue that supporting stateless callers is a goal for Nexus and I would vote to build this functionality into the protocol. 
5. Names aren't final, I'll finalize them when I do the API proposal.

I'll wait for a round of responses here and update the proposal based on the feedback.

Reply:
> Accept optional caller identifiers and namespace the caller and tell caller that their identifier must be unique for their namespace

Would those IDs be used to reference async operations in the cancel, describe, and get result APIs? Do they replace handler provided reference IDs? If that's your intention, the handler will likely need a persistent mapping from client ID to some addressable resource because it doesn't know which operation an ID is referring to and if it's backed by a workflow/activity/update. With the options I provided, the handler can statelessly map client input to a reference ID.

> GetReferenceID should not be part of Nexus

I think we want some stateless method of mapping client input to a handler resource.

Reply:
> Would those IDs be used to reference async operations in the cancel, describe, and get result APIs?

Yes. Same as workflow IDs.

> Do they replace handler provided reference IDs?

They are the same. The handler should only provide an ID if it is not given one.

> If that's your intention, the handler will likely need a persistent mapping from client ID to some addressable resource because it doesn't know which operation an ID is referring to and if it's backed by a workflow/activity/update.

Depends on the situation. Maybe. But if I were a handler, I'd use client's namespace + client-provided ID tuple as the workflow ID and not use a mapping. If it wasn't for workflow ID but instead for, say update on a workflow, my handler is smart enough to know what each thing is for (maybe I prepend the word "update" or something).

I think I understand your issue now. Yes, I would say that every call on an "operation" is qualified by its service name and operation. You don't need the separate by ID. The way I read it earlier, when I saw `Cancel(operation, id)` I thought you were burdening the user with operation, but a lot of the time a user will use a handle which will have that information.

So now that I have my head right, I think of "operations" as REST resources. Just having the resource identifier is not enough, you need the resource name too. I think we should call it "operation name". So technically it's more like:

* `Start(namespace, service name, operation name, (optional) ID, input, ...)`
* `Cancel(namespace, service name, operation name, ID)`
* `GetResult(namespace, service name, operation name, ID)` + `Describe(namespace, service name, operation name, ID)`
  * I think these are probably the same thing, but whatever is decided, I shouldn't have to make two calls to "get status or result"

Sorry for my confusion. I am on team client-provided-optional-identifier + operation name on every call.

Reply:
> `Start(namespace, service name, operation name, (optional) ID, input, ...)`

What's `namespace` here? There's no concept of namespaces in Nexus.

I would rather avoid requiring the client to concatenate multiple identifiers into a single string (and the handler parse and extract those identifiers) when an operation is backed by a resource that is addressed using a complex identifier.
The last API I suggested in this thread avoids this issue.

The only reason I'd want the handler to return a reference id is so Temporal (or any other platform) performing calls on behalf of the user wouldn't need to understand what part of the input is an identifier.

The other alternative might be something like @rylandg mentioned, where part of the input is provided in a dedicated structured identifier field. But I think we want client identifiers to be optional and don't want to burden every client (especially stateful ones, e.g. workflows) to care about what the handler considers an identifier. This approach would also make coding the handlers annoying, as they would have to extract the input from multiple fields.

Reply:
If the identifier is optional, how does temporal call the cancel and describe apis on behalf of the workflow?

Reply:
Because the identifier is always returned, it's only optional for the client to provide. The handler is expected to make the identifier if the client doesn't provide it. The handler must always return an identifier for an operation it starts. Sorry I wasn't clear. Identifier is only optional from a client POV.

Reply:
So what happens if the client does not provide an id, and crashes after performing the operation but before seeing the reply. Do we have an "orphan" operation that nobody can cancel or describe?

Reply:
@cretz I think we (almost) want the same thing then.
I just prefer to not have the client id encoded in the protocol and no string concatenation and parsing, that’s why I’m suggesting that we substitute the string id with structured input.

Reply:
I also have the http use case in mind (see the proposal and the reference to http paths), urls can be mapped to structured input fairly easily. I would want an http gateway to be responsible for this mapping

Reply:
>  I just prefer to not have the client id encoded in the protocol and no string concatenation and parsing, that’s why I’m suggesting that we substitute the string id with structured input.

I completely agree wrt structured input. I think you need 4 structured things to uniquely identify an operation: some kind of client/caller-level namespacing, service name, operation name, operation ID (but this can be optional on client start, but required on handler return and all other operation identification uses).

Reply:
I'll clarify my comment above so we'll see how far off we are.
In the **start** call, the caller sends arbitrary input to the handler, which may contain identifiers, but there is no string ID field on that call.

Here's a more condensed version of my proposal above:

```
Start(service, operation, input, ...) -> reference_id
Cancel(service, reference_id | structured_id)
GetResult(service, reference_id | structured_id)
Describe(service, reference_id | structured_id)
```

The `structured_id` field can take any payload-like form. Here are some examples values:
```
"some-string-identifier"
{ "customerId": "...", "transactionId": "..." }
```

The start call may optionally include a structured identifier as part of the input, such as:
```
Start("some-service", "some-op", { "id": "some-string-identifier", "arg1": 3 })
Start("payments", "charge", { "customerId": "foo", "transactionId": "bar", "amount": 50 })
```

If the caller was responsible for generating complex string identifiers the call above would look something like this:
```
Start("payments", "charge", "foo/bar", { "amount": 50 })
```

When we expose HTTP gateways on top of Nexus, this complex identifier is likely going to be part of the path in the form of: `/payments/charges/customers/foo/charges/bar`, but I don't this string mess in the basic Nexus protocol.

Again, IDs are usually simple values, but there are cases where they are complex and require multiple fields to address a specific resource on the handler side. A stateless client would still be required to provide all of these fields so a handler can address a specific resource but there shouldn't need to be any string encoding or parsing.
Stateful clients (e.g. workflows), do not have to deal with identifiers at all, Temporal takes care of that under the hood, since it can use the handler returned `reference_id`.

Reply:
I'll just add that if the handler wants to hide complex addressability details from stateless clients, it can use a persistent mapping of client ID to internal resource, allowing the client to use simple identifiers in the `Cancel`, `Describe`, and `GetResult` APIs.

I have changed my mind since originally writing this proposal, and I don't think Temporal should provide this persistent mapping capability, at least not initially.

Reply:
I think this is overcomplicating it. We don't need some arbitrary identifier that can be any JSON object of any size. I think there should be one ID to reference an operation and that's it. Very simple. Here's how I would represent your API (assuming namespace was implicit somewhere):

```
Start(service, operation, (optional) operation_id, input, ...) -> operation_id
Cancel(service, operation, operation_id)
GetResult(service, operation, operation_id)
Describe(service, operation, operation_id)
```

Start without ID:

```
Start(service="my-service", operation="my-operation", input={"some-key": "some-val"})
```

Start with ID:

```
Start(service="my-service", operation="my-operation", operation_id="my-operation-id", input={"some-key": "some-val"})
```

That simple. We really shouldn't make this some complex thing. Also, this translates to HTTP well.

> When we expose HTTP gateways on top of Nexus, this complex identifier

Just get rid of the notion of a complex identifier. Get rid of the idea of multiple ways to refer to an operation. Keep it simple.

> handler wants to hide complex addressability

Just get rid of the notion of a complex addressability. Get rid of the idea of multiple ways to refer to an operation. Keep it simple.

> I don't think Temporal should provide this persistent mapping capability, at least not initially.

I'm kinda with you, but I think there should be metadata. Every handler call for an operation should be able to return metadata (i.e. memo) outbound and get it back inbound. They could use this for mapping if they wanted.

Reply:
Same as it would for any other operation (if they wanted to provide an ID). Why does the client need to know it's a workflow update? An operation is an operation to be identified.

Now, if you're asking "how does the handler know what workflow the update needs to go to", that is an operation-specific piece so that is operation-specific parameters. If I were a handler that needed to write an user-facing operation that interacted with another user-facing operation, I would ask the user to provide me as part of input the identifier of that other operation to be interacted with. But if I were a handler that needed to write a user-facing operation that interacted with an internal workflow, then I would require the user as part of input to give whatever was needed to address that workflow.

An operation identifier is a simple concept. Operations that needs to address _other things_ require that the user provide what is needed to address those other things as parameters to my operation. But it sure isn't part of the Nexus protocol.

Reply:
I buy that ids are simple for the client side, and TBH what you're proposing is almost identical to my last option except that you restrict an ID to be a single string which is separate from the rest of the input.

If we go with:
```
Start(service, operation, (optional) operation_id, input, ...) -> operation_id
Cancel(service, operation, operation_id)
GetResult(service, operation, operation_id)
Describe(service, operation, operation_id)
```

There's a few things to note:
1. There's now a single concept of an operation ID, there's no structured way to differentiate between caller and handler generated IDs, requiring the handler to qualify all IDs as if they are client generated.
2. With a handler generated ID, the `Cancel`, `Describe` and `GetResult` calls do not need an operation name in addition to the ID to qualify an operation. This has some benefits.
3. We may want to provide the metadata store for operations with Temporal to avoid burdening callers with handler resource addressable identifiers, this could come at a later stage.
4. Caller/handler may agree on some encoded ID format so callers can directly reference handler resources to avoid the overhead of persistent metadata lookup per operation. We may want to document this.

Overall, I think this API works well.

Reply:
So the caller lib could be something like this, with service provided at client creation and operation/id in a handle?

```ts
const client = await NexusClient.connect({ 
  url: '<url>', 
  service: 'loren-staging' 
})

const handle = await client.start({ 
  operation: 'charge',  
  operationId: transactionId,
  input: { customerId: 'someCustomerId', amount: 5000 },
  callback: 'nexus://caller.example.com/callback?token=' + token
})

const description = await handle.describe()
await handle.cancel()  
const result = await handle.getResult()  
```

Benefit of not requiring `operationId` is that caller doesn't have to do the work of creating it? If caller doesn't create it, they should have an idempotency key in the input? Is idempotency key much easier to create than an operationId? If yes, then the only benefit is they have the ease of providing neither operationId nor idempotency key (and thus risking multiple operations, which may be a fine risk for their use case)?

Reply:
> Benefit of not requiring operationId is that caller doesn't have to do the work of creating it?

Yes. Many calls they just care about the created operation ID. This becomes a lot clearer when looking at a potential HTTP `POST` impl of this which allows one to leave off the ID of the URL.

> If caller doesn't create it, they should have an idempotency key in the input?

There is no requirement for whole-operation idempotency I don't believe. But if there is an operation for _this request_ idempotency to support client retries, that I think is a separate concern.